File: C/Numerical Differentiation/Second-Order Central Difference/centralDifference.c
#include <stdio.h>

/**
 * Implements Second-Order Central Difference for the first derivative.
 * 
 * @param f - Function pointer for the function to differentiate
 * @param x - Point at which to compute the derivative
 * @param h - Step size
 * @return Approximated derivative at x
 */
double central_difference(double (*f)(double), double x, double h) {
    // Formula: (f(x + h) - f(x - h)) / (2 * h)
    return (f(x + h) - f(x - h)) / (2.0 * h);
}

// --- Test the function ---

// Define the function: f(x) = x^2
double f_example(double x) {
    return x * x;
}

int main() {
    double x = 2.0;
    double h = 0.01;
    
    // Execute the calculation
    double df = central_difference(f_example, x, h);
    
    // Print the result
    printf("Central Difference Result: %.5f\n", df);
    
    return 0;
}

File: Python/Numerical Differentiation/Second-Order Central Difference/centralDifference.py
def central_difference(f, x, h):
    """
    Implements Second-Order Central Difference for the first derivative.
    
    Inputs:
        f - Function handle (callable)
        x - Point at which to compute the derivative
        h - Step size
        
    Output:
        df - Approximated derivative at x
    """
    # Formula: (f(x + h) - f(x - h)) / (2 * h)
    df = (f(x + h) - f(x - h)) / (2 * h)
    return df

# --- Test the function ---
if __name__ == "__main__":
    # Define the function: f(x) = x^2
    f = lambda x: x**2
    
    # Define point and step size
    x = 2
    h = 0.01
    
    # Calculate derivative
    df = central_difference(f, x, h)
    
    # Print result
    print(f"Central Difference Result: {df:.5f}")
    # Exact derivative of x^2 at x=2 is 4.0. 
    # For a quadratic function, central difference is often very precise.